### PR TITLE
Fix unnecessary toList/fromList calls during encode/decode process

### PR DESCRIPTION
**Pull request by tarrinneal**
Fix unnecessary toList/fromList calls during encode/decode process.

also makes some kotlin code more idiomatic.

removes the ability to have collisions with the name `list`.

fixes https://github.com/flutter/flutter/issues/119351

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by stuartmorgan** [comment]
> removes double encode/decode on custom classes.

Per offline discussion I think we should adjust the PR description here, as well as the changelog. It's not that things were double-encoding custom classes, it's that we were making the encoding logic needlessly complex by calling ToList/FromList directly instead of letting the encoder do the encoding internally.

(This PR will actually make the byte encoding slightly *less* efficient; before a custom class was encoded as `<list type marker><list size><list elements>` and now it will be `<custom class type marker><list type marker><list size><list elements>`. Maybe that's why it was originally written the other way, but I don't think that's a worthwhile optimization for the conceptual complexity it adds.)

File: packages/pigeon/lib/objc_generator.dart [hunk 1]
```diff
@@ -1488,21 +1488,11 @@ String _makeObjcSignature({
 void generateObjcHeader(ObjcOptions options, Root root, Indent indent) {}
 
 String _listGetter(String list, NamedType field, int index, String? prefix) {
```
Comment (stuartmorgan) [inline packages/pigeon/lib/objc_generator.dart:1490]: This is only used in one place; now that it's one line we should just inline it at the call site. The string itself is pretty self-explanatory.

File: packages/pigeon/lib/swift_generator.dart [hunk 2]
```diff
@@ -214,7 +214,7 @@ class SwiftGenerator extends StructuredGenerator<SwiftOptions> {
           final String fieldName = field.name;
           final String nullsafe = field.type.isNullable ? '?' : '';
           if (field.type.isClass) {
```
Comment (stuartmorgan) [inline packages/pigeon/lib/swift_generator.dart:216]: This is now identical to the final `else`, just written slightly differently; they should be consolidated and normalized (either with everything using `fieldName`, or with `fieldName` eliminated in favor of `field.name`).

File: packages/pigeon/lib/swift_generator.dart [hunk 3]
```diff
@@ -576,11 +578,10 @@ class SwiftGenerator extends StructuredGenerator<SwiftOptions> {
...
       if (type.isClass) {
         indent.writeln('var $variableName: $fieldType? = nil');
-        indent
-            .write('if let ${variableName}List: [Any?] = nilOrValue($value) ');
+        indent.write(
+            'if let ${variableName}List: $fieldType = nilOrValue($value) ');
         indent.addScoped('{', '}', () {
-          indent.writeln(
-              '$variableName = $fieldType.fromList(${variableName}List)');
+          indent.writeln('$variableName = ${variableName}List as $fieldType');
```
Comment (stuartmorgan) [inline packages/pigeon/lib/swift_generator.dart:584]: Shouldn't these no longer have `List` in the name?

(Also, is this still actually different than the `else` branch?)

File: packages/pigeon/lib/swift_generator.dart [hunk 4]
```diff
@@ -603,8 +604,7 @@ class SwiftGenerator extends StructuredGenerator<SwiftOptions> {
       }
     } else {
       if (type.isClass) {
-        indent.writeln(
-            'let $variableName = $fieldType.fromList($value as! [Any?])!');
+        indent.writeln('let $variableName = $value as! $fieldType');
```
Comment (stuartmorgan) [inline packages/pigeon/lib/swift_generator.dart:607]: Is this different than the `else`?

File: packages/pigeon/pigeons/core_tests.dart [hunk 5]
```diff
@@ -39,7 +39,7 @@ class AllTypes {
   Int64List a8ByteArray;
   Float64List aFloatArray;
   // ignore: always_specify_types, strict_raw_type
-  List aList;
+  List list;
```
Comment (stuartmorgan) [inline packages/pigeon/pigeons/core_tests.dart:42]: I think it's worth commenting on why this name doesn't match all the other ones, so someone doesn't think it's a mistake a change it back later.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/pigeon/example/app/android/app/src/main/kotlin/dev/flutter/pigeon_example_app/Messages.g.kt [hunk 6]
```diff
@@ -3,6 +3,7 @@
 // found in the LICENSE file.
 // Autogenerated from Pigeon, do not edit directly.
 // See also: https://pub.dev/packages/pigeon
+@file:Suppress("UNCHECKED_CAST", "LocalVariableName", "ArrayInDataClass")
```
[the remark this answers is not in the document]
Comment (tarrinneal) [reply packages/pigeon/example/app/android/app/src/main/kotlin/dev/flutter/pigeon_example_app/Messages.g.kt:6]: most of them are full file regardless, it gives "redundant suppression" errors.

**Comment by stuartmorgan** [comment]
Now I remember why this error looked familiar: it's a bug in VS 2019, which was fixed in VS 2022.

We actually shouldn't still be using VS 2019 in CI, but it's going to be tricky to update that right now so I'll dig up the workaround. IIRC it will just require turning on exceptions for the test app.

**Comment by tarrinneal** [comment]
> Now I remember why this error looked familiar: it's [a bug in VS 2019, which was fixed in VS 2022](https://developercommunity.visualstudio.com/t/stdany-doesnt-link-when-exceptions-are-disabled/376072).
> 
> We actually shouldn't still be using VS 2019 in CI, but it's going to be tricky to update that right now so I'll dig up the workaround. IIRC it will just require turning on exceptions for the test app.

Seems like it may be something else still...

**Comment by stuartmorgan** [comment]
> > Now I remember why this error looked familiar: it's [a bug in VS 2019, which was fixed in VS 2022](https://developercommunity.visualstudio.com/t/stdany-doesnt-link-when-exceptions-are-disabled/376072).
> > We actually shouldn't still be using VS 2019 in CI, but it's going to be tricky to update that right now so I'll dig up the workaround. IIRC it will just require turning on exceptions for the test app.
> 
> Seems like it may be something else still...

Or it is that same problem, but I put the workaround on the example app binary when the failure is coming from a unit test file that's not compiled in that binary, and I need to change the build settings of the thing that was actually failing to compile 🤦🏻

File: packages/pigeon/lib/swift_generator.dart [hunk 7]
```diff
@@ -969,6 +945,9 @@ String _swiftTypeForDartType(TypeDeclaration type) {
 
 String _nullsafeSwiftTypeForDartType(TypeDeclaration type) {
   final String nullSafe = type.isNullable ? '?' : '';
+  if (type.baseName == 'Map') {
```
Comment (stuartmorgan) [inline packages/pigeon/lib/swift_generator.dart:948]: Maybe a comment in the code here for why Map needs special handling? It's not obvious to me what this change is for.

**Review by stuartmorgan — APPROVED** [review]
LGTM! Just a few minor comments about possible simplifications.

**Pull request merged**